src/t/d14.clj

```diff
@@ -0,0 +1,187 @@
+(ns t.d14
+  (:gen-class)
+  (:require
+    [clojure.string :as cljstr]
+    [clojure.set :as cljset]
+    [clojure.core.match :refer [match]]))
+
+(def sample
+"p=0,4 v=3,-3
+p=6,3 v=-1,-3
+p=10,3 v=-1,2
+p=2,0 v=2,-1
+p=0,0 v=1,3
+p=3,0 v=-2,-2
+p=7,6 v=-1,-3
+p=3,0 v=-1,-2
+p=9,3 v=2,3
+p=7,3 v=-1,2
+p=2,4 v=2,-3
+p=9,5 v=-3,-3
+")
+
+(defn parse-input
+  [input]
+  (->> input
+       (re-seq #"p=(\d+),(\d+) v=(-?\d+),(-?\d+)")
+       (map (fn [[_ & r]] (map parse-long r) ))))
+
+(defn teleport-1d
+  [outsidex xmax]
+  ;; x and y are not necessarily both outside
+  ;; can treat directions separatly
+  (cond
+    (>= outsidex xmax) (- outsidex xmax)
+    (< outsidex 0) (+ xmax outsidex)
+    :else outsidex))
+
+(defn max-steps-1d
+  [x dx xmax]
+  (if (pos? dx)
+    (quot (- (dec xmax) x) dx)
+    (quot x (abs dx))
+    )
+  )
+
+(defn steps-and-1tp
+  [x y dx dy xmax ymax steps]
+  (let [max-steps-in (min
+                       (max-steps-1d x dx xmax)
+                       (max-steps-1d y dy ymax))]
+    (if
+      (>= max-steps-in steps) [(+ x (* dx steps)) (+ y (* dy steps)) 0]
+      (let [x-outside (+ x (* dx (inc max-steps-in)))
+            y-outside (+ y (* dy (inc max-steps-in)))]
+        [(teleport-1d x-outside xmax) (teleport-1d y-outside ymax) (- steps max-steps-in 1)]))))
+
+(defn all-steps
+  [x y dx dy xmax ymax Nsteps]
+  (loop [steps Nsteps
+         lx x
+         ly y]
+    (if (zero? steps)
+      [lx ly]
+      (let [[nx ny nsteps] (steps-and-1tp lx ly dx dy xmax ymax steps)]
+;;        (prn [:prev [:pos lx ly :steps steps :dir dx dy]])
+;;        (prn [:next [:pos nx ny :steps nsteps :dir dx dy]])
+        (recur nsteps nx ny)))))
```

Comment:
I'm a bit confused by this code. I get that you forgot about `mod`, but why split the movement like this? Why not just call `teleport-1d` after each step?

I may be missing something, but how is your code semantically different from this untested snippet?

```clojure
(defn all-steps
  [x y dx dy xmax ymax steps]
  (loop [steps steps, x x, y y]
    (if (zero? steps)
      [x y]
      (recur (dec steps) (teleport-1d (+ x dx) xmax) (teleport-1d (+ y dy) ymax)))))
```

Reply:
because I am not doing one step at a time, but as many as I can before teleporting. So: first I move to the edge of the room, then I TP, then I go back to my main loop and do so again until I have used all my steps. (optimizing when it's not necessary for sure)

For part 2 I define a "one-step" function that does what you suggest (still did not think of mod, I am starting to see why I am spending way too much time on those puzzles T.T )

Reply:
Ah, if that was done for performance reasons, I think it may indeed be a bit misguided. Maybe we should try and measure it.

Reply:
It was, I expected a "this map is 1000 times larger" or something of the kind for the next part ...  it makes no sense when you consider the modulo function but if the TP was different it might have been a good enough thing ??
We could measure, I haven't done that much but it could be fun. (do we have the time?)

Reply:
> (do we have the time?)

Probably not. 🙁